common/src/main/java/com/revenuecat/purchases/common/offlineentitlements/OfflineEntitlementsManager.kt

```diff
@@ -1,18 +1,100 @@
+
+    fun resetOfflineCustomerInfoCache() {
+        synchronized(this) {
+            if (_offlineCustomerInfo != null) {
+                warnLog(OfflineEntitlementsStrings.RESETTING_OFFLINE_CUSTOMER_INFO_CACHE)
+                _offlineCustomerInfo = null
+            }
+        }
+    }
+
+    fun shouldCalculateOfflineCustomerInfoInGetCustomerInfoRequest(
+        isServerError: Boolean,
+        appUserId: String
+    ): Boolean {
+        return appConfig.areOfflineEntitlementsEnabled &&
+            isServerError &&
+            deviceCache.getCachedCustomerInfo(appUserId) == null
+    }
+
+    fun shouldCalculateOfflineCustomerInfoInPostReceipt(isServerError: Boolean): Boolean {
+        return appConfig.areOfflineEntitlementsEnabled &&
+            isServerError
+    }
+
+    @Suppress("FunctionOnlyReturningConstant")
+    fun calculateAndCacheOfflineCustomerInfo(
+        appUserId: String,
+        onSuccess: (CustomerInfo) -> Unit,
+        onError: (PurchasesError) -> Unit
+    ) {
+        if (!appConfig.areOfflineEntitlementsEnabled) {
+            onError(PurchasesError(
+                PurchasesErrorCode.UnsupportedError,
+                OfflineEntitlementsStrings.OFFLINE_ENTITLEMENTS_NOT_SUPPORTED
+            ))
+            return
+        }
+        synchronized(this) {
+            val alreadyProcessing = offlineCustomerInfoCallbackCache.containsKey(appUserId)
+            val callbacks = offlineCustomerInfoCallbackCache[appUserId] ?: emptyList()
+            offlineCustomerInfoCallbackCache[appUserId] = callbacks + listOf(onSuccess to onError)
+            if (alreadyProcessing) {
+                debugLog(OfflineEntitlementsStrings.ALREADY_CALCULATING_OFFLINE_CUSTOMER_INFO.format(appUserId))
+                return
+            }
+        }
+        offlineCustomerInfoCalculator.computeOfflineCustomerInfo(
+            appUserId,
+            onSuccess = { customerInfo ->
+                synchronized(this) {
+                    warnLog("Updating offline customer info cache")
+                    _offlineCustomerInfo = customerInfo
+                    val callbacks = offlineCustomerInfoCallbackCache.remove(appUserId)
+                    callbacks?.forEach { (onSuccess, _) ->
+                        onSuccess(customerInfo)
+                    }
+                }
+            },
+            onError = {
+                synchronized(this) {
+                    val callbacks = offlineCustomerInfoCallbackCache.remove(appUserId)
+                    callbacks?.forEach { (_, onError) ->
+                        onError(it)
+                    }
+                }
+            }
+        )
+    }
 
     fun updateProductEntitlementMappingCacheIfStale() {
-        if (deviceCache.isProductEntitlementMappingCacheStale()) {
+        if (appConfig.areOfflineEntitlementsEnabled && deviceCache.isProductEntitlementMappingCacheStale()) {
```

Comment:
I added the feature flag for updating the mapping as well, so we don't need to keep that code commented out.